File: src/Scripting/Lua/prototype.lua
local ServiceProfile = {}
ServiceProfile.__index = ServiceProfile

function ServiceProfile.new(name, features)
    return setmetatable({ name = name, features = features }, ServiceProfile)
end

function ServiceProfile:clone()
    local features = {}
    for i, feature in ipairs(self.features) do
        features[i] = feature
    end
    return ServiceProfile.new(self.name, features)
end

function ServiceProfile:describe()
    return self.name .. ": " .. table.concat(self.features, ",")
end

local original = ServiceProfile.new("orders", { "metrics" })
local canary = original:clone()
canary.name = "orders-canary"
table.insert(canary.features, "tracing")

print("original=" .. original:describe())
print("clone=" .. canary:describe())

File: src/Scripting/PHP/prototype.php
<?php

declare(strict_types=1);

interface Prototype
{
    public function copy(): static;
}

final class ServiceProfile implements Prototype
{
    /** @param list<string> $features */
    public function __construct(
        public string $name,
        public array $features,
    ) {
    }

    public function copy(): static
    {
        return clone $this;
    }

    public function describe(): string
    {
        return $this->name . ': ' . implode(',', $this->features);
    }
}

$original = new ServiceProfile('orders', ['metrics']);
$canary = $original->copy();
$canary->name = 'orders-canary';
$canary->features[] = 'tracing';

echo 'original=' . $original->describe() . PHP_EOL;
echo 'clone=' . $canary->describe() . PHP_EOL;
